### PR TITLE
cmd/pprofessor: introduce pprof-based tool

### LINKED ISSUE #3828: pprof-based tool for visualising aggregated profiles
Create a customised [pprof](https://github.com/google/pprof) for visualising profiles stored by apm-server in Elasticsearch. The tool will fetch data directly from Elasticsearch using aggregations, and visualise using the standard pprof UI.

This would not be part of the product, it would be for development and testing of apm-server itself.

### PR DESCRIPTION
**Pull request by axw**
## Motivation/summary

pprofessor is based on the pprof tool, using an alternative
"Fetcher" implementation that queries Elasticsearch to
aggregate continuous profiling samples recorded by APM Server.

I haven't gone through all of the pprof flags to make sure
they work, so some may not make sense. There are a couple of
flags added specifically for our use case:

* `-service`, which defines the service name to filter on.
* `-start`, which defines the start timestamp (or date math)
  from which to start aggregating. If specified along with
  `-seconds`, the latter  controls the end time. If only `-start`
  is specified, then the end time is now. If only `-seconds` is
  specified, then the end time is now and the start time is duration
  seconds before now.

We record the number of profiles and documents aggregated in
profile comments. All metrics (currently: cpu, inuse/allocated
heap objects/space) are aggregated into a single profile.

To aggregate duration correctly, we introduce the `profile.id`
field, which is a unique ID generated per profile, shared by
all sample docs derived from a profile.

No automated tests for now, and comes with no warranty or support.

Note that there is a known issue with how the Go Agent reports "alloc"
samples: it reports ever-increasing counters, rather than deltas. We
need to change this, so we can visualise the allocations within a define
time range. See https://github.com/elastic/apm-agent-go/issues/708

## Checklist

- [x] I have signed the [Contributor License Agreement](https://www.elastic.co/contributor-agreement/).
~- [ ] I have updated [CHANGELOG.asciidoc](https://github.com/elastic/apm-server/blob/master/CHANGELOG.asciidoc)~

I have considered changes for:
~- [ ] documentation~
~- [ ] logging (add log lines, choose appropriate log selector, etc.)~
~- [ ] metrics and monitoring (create issue for Kibana team to add metrics to visualizations, e.g. [Kibana#44001](https://github.com/elastic/kibana/issues/44001))~
- [ ] automated tests (add tests for the code changes, all [**unit** tests](https://github.com/elastic/apm-server/blob/master/TESTING.md) pass locally)
~- [ ] telemetry~
~- [ ] Elasticsearch Service (https://cloud.elastic.co)~
~- [ ] Elastic Cloud Enterprise (https://www.elastic.co/products/ece)~
~- [ ] Elastic Cloud on Kubernetes (https://www.elastic.co/elastic-cloud-kubernetes)~

## How to test these changes

- Run APM Server with CPU and Heap profiling enabled in APM Server:

`apm-server -E apm-server.instrumentation.enabled=true -E apm-server.instrumentation.profiling.cpu.enabled=true -E apm-server.instrumentation.profiling.cpu.interval=10s -E apm-server.instrumentation.profiling.heap.enabled=true`

- Run pprofessor:

`go run ./cmd/pprofessor -start=now-1h -service=apm-server --http=:6060 http://admin:changeme@localhost:9200`

## Demo

pprofessor3

## Related issues

Closes #3828

### CODE REVIEW

**Comment by jalvz** [comment]
This doesn't seem to work for me with the 7.8 stack:

```http://admin:changeme@localhost:9200: [400 Bad Request] {"error":{"root_cause":[{"type":"x_content_parse_exception","reason":"[1:103] [composite] after doesn't support values of type: VALUE_NULL"}],"type":"x_content_parse_exception","reason":"[1:103] [composite] after doesn't support values of type: VALUE_NULL"},"status":400}```

A few other notes/questions:

- Non obvious requirements for running this should be documented (min stack version, Graphviz, etc) 
- The command semantics are not so intuitive to me: 
  Since this is mostly for us, I'd expect `service` to be hardcoded to `apm-server`, or at least default to it. 
  The behaviour of specifying only `seconds` is conventional?  it doesn't look so useful..
  I think the Elasticsearch URL should have its own flag, as other required arguments have it too.  
- While not tested, I'd remove all non-essential code, like  tls_ca handling.   
- The expression "unique profile ID shared by all samples" is not so intuitive, consider `profileKey`, `profileGroupID`, or similar.
- It would be more coherent to use types for both Elasticsearch responses and queries or none, but not only responses.
- I think Elasticsearch types should live on their own package, or at least their own file.

**Comment by axw** [comment]
@jalvz thanks for the review!

> http://admin:changeme@localhost:9200: [400 Bad Request] {"error":{"root_cause":[{"type":"x_content_parse_exception","reason":"[1:103] [composite] after doesn't support values of type: VALUE_NULL"}],"type":"x_content_parse_exception","reason":"[1:103] [composite] after doesn't support values of type: VALUE_NULL"},"status":400}

That's probably because there's a composite aggregation on `profile.id`, which is introduced in this PR. You would need to build and run apm-server to use the tool. (Part of the reason for creating this tool is to surface missed data/requirements such as this.)

> Non obvious requirements for running this should be documented (min stack version, Graphviz, etc)

I can add a basic README. Until then: my intention was for the tool to be used with the same (branch) stack & server version. Otherwise, no additional dependencies apart from pprof's (which includes Graphviz).

> The command semantics are not so intuitive to me:
> Since this is mostly for us, I'd expect service to be hardcoded to apm-server, or at least default to it.

I'll set "apm-server" as the default.

> The behaviour of specifying only seconds is conventional? it doesn't look so useful..

That's one of the standard `pprof` flags. If you run `go tool pprof -seconds=60 http://service:1234/debug/pprof/profile` it'll profile for 60 seconds.

> I think the Elasticsearch URL should have its own flag, as other required arguments have it too.

This is part of the pprof tool: the "source" (i.e. Elasticsearch URL in this case) is a positional argument.

> While not tested, I'd remove all non-essential code, like tls_ca handling.

Sure, I'll remove it until it's needed.

> The expression "unique profile ID shared by all samples" is not so intuitive, consider profileKey, profileGroupID, or similar.

Are you referring to the wording in `fields.yml`? I'll try to clarify the description if that's what you mean. I don't think we should change the field name. It is the ID of the profile to which the samples belong. Parent/child (1:n) relationship.

> It would be more coherent to use types for both Elasticsearch responses and queries or none, but not only responses.

The reason for having struct types for responses is because we need to operate on the results. The requests are just JSON marshalled, so it doesn't really matter if they're structs or maps. I did what's simplest.

> I think Elasticsearch types should live on their own package, or at least their own file.

-1 on separate package, see https://dave.cheney.net/practical-go/presentations/qcon-china.html#_consider_fewer_larger_packages

If you think it's really worthwhile I can create a separate file, but I don't think it is. The types are used in exactly one place.

File: cmd/pprofessor/fetcher.go [hunk 1]
```diff
@@ -0,0 +1,442 @@
...
+	flags.AddExtraUsage(fetcherUsage)
+	return &fetcher{
+		flags:       flags,
+		startTime:   flags.String("start", "", "Start time for profile aggregation"),
+		serviceName: flags.String("service", "", "Profiled service name"),
+	}
+}
+
+// Fetch fetches a profile from Elasticsearch, by aggregating samples for a specified service name and time range.
+func (f *fetcher) Fetch(src string, duration, timeout time.Duration) (*profile.Profile, string, error) {
```
[the remark this answers is not in the document]
Comment (axw) [reply cmd/pprofessor/fetcher.go]: Done

**Comment by jalvz** [comment]
> You would need to build and run apm-server to use the tool

Don't follow. Did `make` and `make update`, then ran with relevant flags. Anything else needed?

> no additional dependencies apart from pprof's (which includes Graphviz).

Hmm, I actually had to apt-install it...

**Comment by axw** [comment]
>> You would need to build and run apm-server to use the tool
>
> Don't follow. Did make and make update, then ran with relevant flags. Anything else needed?

Looking at the error message again, maybe I introduced a bug in between testing and proposing this PR. I'll look into it and come back.

>> no additional dependencies apart from pprof's (which includes Graphviz).
>
> Hmm, I actually had to apt-install it...

Sorry, what I meant was "no additional dependencies other than pprof's dependencies, among which is Graphviz" -- not that it's included within the pprof binary.

**Comment by axw** [comment]
> Looking at the error message again, maybe I introduced a bug in between testing and proposing this PR. I'll look into it and come back.

I hadn't tested under the scenario where there was an existing index/template without the `profile.id` field. If you remove your index and overwrite the template then it should work. Either way, I've added a bug fix which will avoid the error noted.

I've added a README and (hopefully) clarified the meaning of `profile.id`.

**Comment by axw** [comment]
Merging without a changelog entry, as this is not intended for end users.

**Pull request merged**